### ISSUE TITLE
[MDEP-579] Regression: get goal does not pass server credentials to BasicRepositoryConnector

### ISSUE THREAD

**Issue by jira-importer** [apache/maven-dependency-plugin#1013]
**[Richard W. Eggert II](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=reggert1980)** opened **[MDEP-579](https://issues.apache.org/jira/browse/MDEP-579?redirect=false)** and commented

The `get` goal does not pass the server credentials from `settings.xml` to the `BasicRepositoryConnector` in version 3.0.1 (and, presumably 3.0.0), resulting in `NotAuthorized` errors when resolving artifacts against repositories that require authentication. It works correctly in version 2.9.

Background: I discovered this in the course of debugging a Jenkins job in which I'm using the `get` and `copy` goals from the command line (with no POM) to download artifacts to deploy. After spending several hours thinking that Jenkins was not properly configuring `settings.xml`, I noticed from the Maven debug output that the credentials were being passed when resolving the maven-dependency-plugin and its dependencies, but not being passed when resolving the artifact I requested. On a hunch I downgraded from maven-dependency-plugin version 3.0.1 to 2.9, and suddenly everything magically worked.

---

**Affects:** 3.0.0, 3.0.1

**Remote Links:**
- [GitHub Pull Request #18
](https://github.com/apache/maven-dependency-plugin/pull/18)
- [GitHub Pull Request #19
](https://github.com/apache/maven-dependency-plugin/pull/19)
- [#23 MRM
](https://github.com/mojohaus/mrm/issues/23)

1 votes, 5 watchers

**Comment by jira-importer** [comment]
**[Pim Moerenhout](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=pmoerenhout)** commented

Hi [Kent Granström](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kentgran), I have made a fix in the following branch:

https://github.com/pmoerenhout/maven-dependency-plugin/tree/MDEP-579

Are you able to check if this resolves the problem ?

**Comment by jira-importer** [comment]
**[Kent Granström](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kentgran)** commented

Hi [Pim Moerenhout](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=pmoerenhout)

I have now cloned your branch and build it and tested it on some artifacts we have got and unfortunatelly it appears not to work. Due to company policy I have had to replace certain data... sorry for that but  this is what I get when I run it: (first the updated version and den the original one)

```

[DEBUG] Configuring mojo org.apache.maven.plugins:maven-dependency-plugin:3.1.2-SNAPSHOT:get from plugin realm ClassRealm[plugin>org.apache.maven.plugins:maven-dependency-plugin:3.1.2-SNAPSHOT, parent: sun.misc.Launcher$AppClassLoader@33909752]
[DEBUG] Configuring mojo 'org.apache.maven.plugins:maven-dependency-plugin:3.1.2-SNAPSHOT:get' with basic configurator -->
[DEBUG] (f) artifact = com.test.note:note-docker:4.1:zip
[DEBUG] (s) packaging = jar
[DEBUG] (f) pomRemoteRepositories = [ id: nexus
 url: https://example.com/nexus/repository/public-repo
 layout: default
snapshots: [enabled => false, update => daily]
 releases: [enabled => true, update => daily]
]
[DEBUG] (f) remoteRepositories = internal-repo-id::default::https://example.com/nexus/repository/internal-repo
[DEBUG] (f) session = org.apache.maven.execution.MavenSession@682bd3c4
[DEBUG] (f) skip = false
[DEBUG] (f) transitive = false
[DEBUG] -- end configuration --
[INFO] Resolving com.test.note:note-docker:zip:4.1
[DEBUG] Using transporter WagonTransporter with priority -1.0 for https://example.com/nexus/repository/public-repo
[DEBUG] Using connector BasicRepositoryConnector with priority 0.0 for https://example.com/nexus/repository/public-repo with username=***, password=***
Downloading: https://example.com/nexus/repository/public-repo/com/test/note/note-docker/4.1/note-docker-4.1.pom
[DEBUG] Writing tracking file C:\Users\abcd\.m2\repository\com\test\note\note-docker\4.1\note-docker-4.1.pom.lastUpdated
[DEBUG] Skipped remote request for com.test.note:note-docker:pom:4.1, already updated during this session.
[WARNING] The POM for com.test.note:note-docker:zip:4.1 is missing, no dependency information available

[DEBUG] Using transporter WagonTransporter with priority -1.0 for https://example.com/nexus/repository/public-repo
[DEBUG] Using connector BasicRepositoryConnector with priority 0.0 for https://example.com/nexus/repository/public-repo with username=***, password=***
Downloading: https://example.com/nexus/repository/public-repo/com/test/note/note-docker/4.1/note-docker-4.1.zip
[DEBUG] Writing tracking file C:\Users\abcd\.m2\repository\com\test\note\note-docker\4.1\note-docker-4.1.zip.lastUpdated
[DEBUG] Skipped remote request for com.test.note:note-docker:zip:4.1, already updated during this session.
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 6.206 s
[INFO] Finished at: 2019-08-12T07:21:20+02:00
[INFO] Final Memory: 17M/207M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-dependency-plugin:3.1.2-SNAPSHOT:get (default-cli) on project dummy-artifact: Couldn't download artifact: Could not find artifact com.test.note:note-docker:zip:4.1 in nexus (https://example.com/nexus/repository/public-repo) -> [Help 1]



ORIGINAL 3.1.1

[INFO] Resolving com.test.note:note-docker:zip:4.1
[DEBUG] Using transporter WagonTransporter with priority -1.0 for https://example.com/nexus/repository/public-repo
[DEBUG] Using connector BasicRepositoryConnector with priority 0.0 for https://example.com/nexus/repository/public-repo with username=***, password=***
Downloading: https://example.com/nexus/repository/public-repo/com/test/note/note-docker/4.1/note-docker-4.1.pom
[DEBUG] Writing tracking file C:\Users\abcd\.m2\repository\com\test\note\note-docker\4.1\note-docker-4.1.pom.lastUpdated

[DEBUG] Using transporter WagonTransporter with priority -1.0 for https://example.com/nexus/repository/internal-repo
[DEBUG] Using connector BasicRepositoryConnector with priority 0.0 for https://example.com/nexus/repository/internal-repo
Downloading: https://example.com/nexus/repository/internal-repo/com/test/note/note-docker/4.1/note-docker-4.1.pom
[DEBUG] Writing tracking file C:\Users\abcd\.m2\repository\com\test\note\note-docker\4.1\note-docker-4.1.pom.lastUpdated
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 5.162 s
[INFO] Finished at: 2019-08-12T07:26:33+02:00
[INFO] Final Memory: 17M/168M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-dependency-plugin:3.1.1:get (default-cli) on project dummy-artifact: Couldn't download artifact: Failed to read artifact descriptor for com.test.note:note-docker:zip:4.1:
Could not transfer artifact com.test.note:note-docker:pom:4.1 from/to internal-repo-id (https://example.com/nexus/repository/internal-repo): Not authorized , ReasonPhrase:Unauthorized. -> [Help 1]
```

From what I gather the difference is that in your case the creds are passed to our repo manager (Nexus) but it is for the public-repo (not the internal-repo that also require creds)

**Comment by jira-importer** [comment]
**[Kent Granström](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kentgran)** commented

Hi [Pim Moerenhout](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=pmoerenhout).

We're actually using a usertoken and pwtoken generated by our Nexus instance but that basically equals to username and pw. I'm not allowed to hand out anything from our environment but for the sake of this issue I have written a somewhat equivalent configuration of what I'm doing to get this behaviour. Obviously the path, user and pw doesn't map to the real world but perhaps its enough for you to be able to reproduce the issue. 

Running a bat-file from the commandline in Windows:

```java
In get.bat (toggling 3.1.1 and 3.1.2-SNAPSHOT):

mvn clean -X org.apache.maven.plugins:maven-dependency-plugin:3.1.2-SNAPSHOT:get ^
 -Dartifact= com.test.note:note-docker:4.1:zip ^
 -DremoteRepositories=internal-repo-id::default::https://example.com/nexus/repository/internal-repo ^
 -Dtransitive=false


This is settings.xml:

<?xml version="1.0"?>

<settings>

 <servers>
   <server>
     <id>nexus</id>
     <username>SomeUser</username>
     <password>APassword</password>
   </server>

   <server>
     <id>internal-repo-id</id>
     <username>SomeUSer</username>
     <password>APAssword</password>
   </server>
 </servers>

 <mirrors>
   <mirror>
     <id>nexus</id>
     <name>Our Public repo group</name>
     <url>https://example.com/nexus/repository/public-repo</url>
     <mirrorOf>*,!internal-repo</mirrorOf>
   </mirror>

   <mirror>
     <id>internal-repo-id</id>
     <name>Our internal artifact group</name>
     <url>https://example.com/nexus/repository/internal-repo</url>
     <mirrorOf>internal-repo</mirrorOf>
   </mirror>
 </mirrors>
</settings>
```

Hope this helps.

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

I've added a link to https://github.com/mojohaus/mrm/issues/23 which, once implemented, should make it easier to verify this usecase as part of the integration tests

**Comment by jira-importer** [comment]
**[Pim Moerenhout](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=pmoerenhout)** commented

Hi [Kent Granström](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kentgran),

I have setup the setup you've described, using a Nexus docker image as mirror of public with authentication, together with a private repository. However, I did change the mirrorOf property of the nexus mirror to *,!internal-repo-id, as this property must contain the id. See [Settings reference](https://maven.apache.org/settings.html#Mirrors).

With the setup described, the artifact could be get, from the internal repository, with the correct authentication. Before the get, I delete the artifact from the local repository in the .m2/repo folder.

In the failure log, the message is _Could not find artifact com.test.note:note-docker:zip:4.1 in nexus_. It's not mentioning that the authentication failed.

Could you verify in the real setup how the mirrorOf is specified ?

If it still fails, I could setup the docker image to hold the 2 repositories, as next step.

**Comment by jira-importer** [comment]
**[Kent Granström](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kentgran)** commented

Apologies [Pim Moerenhout](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=pmoerenhout). I made a typo and testing it after that it seems to work just fine. I can now use the "get" and it passes the creds nicely. I would concider this as solved, at least for me :). Perhaps a few testcases ought to be written for this, and similar, scenarios.

Many many thanks [Pim Moerenhout](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=pmoerenhout)

**Comment by jira-importer** [comment]
**[Pim Moerenhout](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=pmoerenhout)** commented

Thanks. I have a testcase ready, using Jetty for the authentication. I will prepare a new pull request.

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

Fixed in [e897911067c643ee891bfc2b05a85f8c7ca658d5](https://gitbox.apache.org/repos/asf?p=maven-dependency-plugin.git;a=commit;h=e897911067c643ee891bfc2b05a85f8c7ca658d5)
Thanks for the patch!
